src/pip/_internal/commands/install.py

```diff
@@ -539,84 +535,122 @@ def _warn_about_conflicts(self, conflict_details):
+    home=None,  # type: Optional[str]
+    root=None,  # type: Optional[str]
+    isolated=False,  # type: bool
+    prefix=None,  # type: Optional[str]
 ):
-    # type:(...) -> List[str]
+    # type: (...) -> List[str]
     scheme = distutils_scheme('', user=user, home=home, root=root,
                               isolated=isolated, prefix=prefix)
     return [scheme['purelib'], scheme['platlib']]
 
 
-def site_packages_writable(root, isolated):
-    # type: (Optional[str], bool) -> bool
-    return all(
-        test_writable_dir(d) for d in set(
-            get_lib_location_guesses(root=root, isolated=isolated))
-    )
+def site_packages_writable(
+    user=False,  # type: bool
+    root=None,  # type: Optional[str]
+    isolated=False,  # type: bool
+    prefix=None,  # type: Optional[str]
+):
+    # type: (...) -> bool
+    return all(map(test_writable_dir, set(get_lib_location_guesses(
+        user=user, root=root, isolated=isolated, prefix=prefix,
+    ))))
 
 
 def decide_user_install(
-    use_user_site,  # type: Optional[bool]
+    use_user_site=None,  # type: Optional[bool]
     prefix_path=None,  # type: Optional[str]
     target_dir=None,  # type: Optional[str]
     root_path=None,  # type: Optional[str]
     isolated_mode=False,  # type: bool
 ):
     # type: (...) -> bool
-    """Determine whether to do a user install based on the input options.
+    """Determine whether to do a user install.
 
-    If use_user_site is False, no additional checks are done.
-    If use_user_site is True, it is checked for compatibility with other
-    options.
-    If use_user_site is None, the default behaviour depends on the environment,
-    which is provided by the other arguments.
-    """
-    # In some cases (config from tox), use_user_site can be set to an integer
-    # rather than a bool, which 'use_user_site is False' wouldn't catch.
-    if (use_user_site is not None) and (not use_user_site):
-        logger.debug("Non-user install by explicit request")
-        return False
+    If use_user_site is None, the behavior is decided upon other
+    arguments and the environment.
 
-    if use_user_site:
-        if prefix_path:
-            raise CommandError(
-                "Can not combine '--user' and '--prefix' as they imply "
-                "different installation locations"
-            )
-        if virtualenv_no_global():
+    Compatibility of installation location options and the environment
+    is also asserted in this function.
+    """
+    # Check for incompatible options
+    locations = {"--target": target_dir is not None, "--user": use_user_site,
+                 "--prefix": prefix_path is not None}
+    destinations = [k for k, v in locations.items() if v]
+    if len(destinations) > 1:
+        last, rest = destinations[-1], ", ".join(destinations[:-1])
+        raise CommandError("Different installation locations are implied by "
+                           "{} and {}.".format(rest, last))
+
+    if target_dir:
```

Comment:
```suggestion
    if locations["--target"]:
```

Likewise for the other checks.

Reply:
mypy doesn't seem to be able to imply from `locations` :disappointed:   I changed the check to be more explicit though.

Reply:
*sigh*